File: unit_tests/primitives/transitions/test_cooperative_hydrolysis.py
#    Copyright (C) 2010 Mark Burnett
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import unittest
from collections import defaultdict

from actin_dynamics.primitives.transitions.cooperative_hydrolysis import *

from actin_dynamics.state.single_strand_filaments import Filament

from unit_tests.mocks.concentrations import MockConcentration


class CooperativeHydrolysisTest(unittest.TestCase):
    def setUp(self):
        self.strand = Filament(['a', 'b', 'c', 'a', 'b', 'c', 'a'])
        self.normal_one = CooperativeHydrolysis(old_state='a',
                rate=3, new_state='b', c=2)
        self.normal_two = CooperativeHydrolysis(old_state='b',
                rate=2, new_state='c', a=3)
        self.normal_three = CooperativeHydrolysis(old_state='a',
                rate=4, new_state='b', b=3)

        self.missing = CooperativeHydrolysis(old_state='d',
                rate=7, new_state=8, e=1.5)

    def test_normal_rates(self):
        self.assertEqual(self.normal_one.R(None, [self.strand],   None), [15])
        self.assertEqual(self.normal_two.R(None, [self.strand],   None), [12])
        self.assertEqual(self.normal_three.R(None, [self.strand], None), [12])

    def test_missing_rates(self):
        self.assertEqual(self.missing.R(None, [self.strand], None), [0])

    def test_perform_normal_boundary(self):
        self.normal_one.perform(None, [self.strand], None, 0, 10)
        self.assertEqual(list(self.strand), ['a', 'b', 'c', 'b', 'b', 'c', 'a'])
        self.assertEqual(self.normal_one.R(None, [self.strand], None), [9])
        self.assertEqual(self.normal_two.R(None, [self.strand], None), [10])

        self.normal_one.perform(None, [self.strand], None, 0, 8)
        self.assertEqual(list(self.strand), ['a', 'b', 'c', 'b', 'b', 'c', 'b'])
        self.assertEqual(self.normal_one.R(None, [self.strand], None), [3])
        self.assertEqual(self.normal_two.R(None, [self.strand], None), [12])

    def test_perform_normal_random(self):
        self.normal_one.perform(None, [self.strand], None, 0, 0)
        self.assertEqual(list(self.strand), ['b', 'b', 'c', 'a', 'b', 'c', 'a'])
        self.assertEqual(self.normal_one.R(None, [self.strand], None), [12])
        self.assertEqual(self.normal_two.R(None, [self.strand], None), [10])

    def test_perform_missing(self):
        self.assertRaises(ZeroDivisionError, self.missing.perform,
                          None, [self.strand], None, 0, 0)


#class CooperativeHydrolysisWithByproductTest(unittest.TestCase):
#    def test_notice(self):
#        # XXX Write some tests for this, and expand the above tests for multi filament.
#        self.assertFalse(True)
#    def setUp(self):
#        self.filament = Filament([1, 2, 3, 1, 2, 3, 1])
#        self.concentrations = defaultdict(MockConcentration)
#
#        self.normal_one = CooperativeHydrolysisWithByproduct(old_state=1,
#                pointed_neighbor=3, rate=3, cooperativity=2,
#                new_state=2, byproduct=11)
#        self.normal_two = CooperativeHydrolysisWithByproduct(old_state=2,
#                pointed_neighbor=1, rate=2, cooperativity=3,
#                new_state=3, byproduct=12)
#        self.normal_three = CooperativeHydrolysisWithByproduct(old_state=1,
#                pointed_neighbor=2, rate=4, cooperativity=3,
#                new_state=2, byproduct=13)
#
#        self.missing = CooperativeHydrolysisWithByproduct(old_state=4,
#                pointed_neighbor=5, rate=7, cooperativity=1.5,
#                new_state=8, byproduct=14)
#
#    def test_normal_rates(self):
#        self.assertEqual(self.normal_one.R(None, self.filament,   None), 15)
#        self.assertEqual(self.normal_two.R(None, self.filament,   None), 12)
#        self.assertEqual(self.normal_three.R(None, self.filament, None), 12)
#
#    def test_missing_rates(self):
#        self.assertEqual(self.missing.R(None, self.filament, None), 0)
#
#    def test_perform_normal_boundary(self):
#        self.normal_one.perform(None, self.filament, self.concentrations, 0)
#        self.assertEqual(self.filament.states, [1, 2, 3, 2, 2, 3, 1])
#        self.assertEqual(self.normal_one.R(None, self.filament, None), 9)
#        self.assertEqual(self.normal_two.R(None, self.filament, None), 10)
#        self.assertEqual(self.concentrations[11].count, 1)
#
#        self.normal_one.perform(None, self.filament, self.concentrations, 0)
#        self.assertEqual(self.filament.states, [1, 2, 3, 2, 2, 3, 2])
#        self.assertEqual(self.normal_one.R(None, self.filament, None), 3)
#        self.assertEqual(self.normal_two.R(None, self.filament, None), 12)
#        self.assertEqual(self.concentrations[11].count, 2)
#
#    def test_perform_normal_random(self):
#        self.normal_one.perform(None, self.filament, self.concentrations, 14)
#        self.assertEqual(self.filament.states, [2, 2, 3, 1, 2, 3, 1])
#        self.assertEqual(self.normal_one.R(None, self.filament, None), 12)
#        self.assertEqual(self.normal_two.R(None, self.filament, None), 10)
#        self.assertEqual(self.concentrations[11].count, 1)
#
#    def test_perform_missing(self):
#        self.assertRaises(IndexError, self.missing.perform,
#                          None, self.filament, self.concentrations, 0)
#        self.assertEqual(self.concentrations[14].count, 0)


if '__main__' == __name__:
    unittest.main()

File: actin_dynamics/stochastic/tests/transitions/test_cooperative_hydrolysis.cpp
//    Copyright (C) 2012 Mark Burnett
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include <gtest/gtest.h>

#include <vector>

#include <boost/assign/std/vector.hpp>

// Bring operator+= into the namespace
using namespace boost::assign;

#include "state.h"

#include "test_states.h"

#include "transitions/cooperative_hydrolysis.h"
#include "filaments/simple_filament.h"
#include "concentrations/fixed_reagent.h"

using namespace stochastic;

class CooperativeHydrolysisTest : public testing::Test {
    protected:
        virtual void SetUp() {
            std::vector<State> values1;
            values1 += zero, one, zero, zero, two, one, zero, one;
            std::vector<State> values2;
            values2 += one, one, two, zero, two, zero, zero, zero;

            filaments.push_back(filaments::Filament::ptr_t(
                        new filaments::SimpleFilament(values1)));
            filaments.push_back(filaments::Filament::ptr_t(
                        new filaments::SimpleFilament(values2)));

            concentrations[zero] = concentrations::Concentration::ptr_t(
                        new concentrations::FixedReagent(0, 1));
            concentrations[one] = concentrations::Concentration::ptr_t(
                        new concentrations::FixedReagent(0, 1));
            concentrations[two] = concentrations::Concentration::ptr_t(
                        new concentrations::FixedReagent(0, 1));
            concentrations[three] = concentrations::Concentration::ptr_t(
                        new concentrations::FixedReagent(0, 1));
            concentrations[four] = concentrations::Concentration::ptr_t(
                        new concentrations::FixedReagent(0, 1));
        }

        virtual void TearDown() {
            filaments.clear();
            concentrations.clear();
        }

        filaments::container_t filaments;
        concentrations::container_t concentrations;
};

TEST_F(CooperativeHydrolysisTest, RandomEquivalence) {
    transitions::CooperativeHydrolysis tr_12(zero, one, two, 3);

    EXPECT_DOUBLE_EQ(15, tr_12.initial_R(0, filaments, concentrations));
    EXPECT_EQ(1, tr_12.perform(0, 14.1, filaments, concentrations));
    EXPECT_DOUBLE_EQ(12, tr_12.R(0, filaments, concentrations, 1));
}

TEST_F(CooperativeHydrolysisTest, Mixed) {
    transitions::CooperativeHydrolysis t01_2(zero, one, two, 3, 11);

    EXPECT_DOUBLE_EQ(75, t01_2.initial_R(0, filaments, concentrations));

    // Random Transition - no effect on boundaries
    EXPECT_EQ(1, t01_2.perform(0, 11.2, filaments, concentrations));
    EXPECT_DOUBLE_EQ(72, t01_2.R(0, filaments, concentrations, 1));

    // Vectorial Transition - subtracts (30 + 3 from Rate)
    EXPECT_EQ(0, t01_2.perform(0, 43, filaments, concentrations));
    EXPECT_DOUBLE_EQ(39, t01_2.R(0, filaments, concentrations, 0));
}

TEST_F(CooperativeHydrolysisTest, MixedWithByproduct) {
    transitions::CooperativeHydrolysisWithByproduct t01_2_3(zero, one, two, 3, three, 11);

    EXPECT_DOUBLE_EQ(75, t01_2_3.initial_R(0, filaments, concentrations));

    // Random Transition - no effect on boundaries
    EXPECT_EQ(1, t01_2_3.perform(0, 11.2, filaments, concentrations));
    EXPECT_DOUBLE_EQ(72, t01_2_3.R(0, filaments, concentrations, 1));
    EXPECT_EQ(1, concentrations[three]->monomer_count());

    // Vectorial Transition - subtracts (30 + 3 from Rate)
    EXPECT_EQ(0, t01_2_3.perform(0, 43, filaments, concentrations));
    EXPECT_DOUBLE_EQ(39, t01_2_3.R(0, filaments, concentrations, 0));
    EXPECT_EQ(2, concentrations[three]->monomer_count());
}
